### PR TITLE
[2.x] fix: concurrency control around Scala 3 compiler

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes https://github.com/sbt/sbt/issues/7873

## Problem
Early initialization thread throws ConcurrentModificationException.

## Solution
This wraps the initialization with Retry(...).

### CODE REVIEW

File: buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala [hunk 1]
```diff
@@ -131,8 +131,10 @@ private[sbt] object SbtParser:
 
   private[sbt] var scalacGlobalInitReporter: Option[ConsoleReporter] = None
 
-  private[sbt] val globalReporter = UniqueParserReporter()
-  private[sbt] val defaultGlobalForParser = ParseDriver()
+  private[sbt] lazy val globalReporter = UniqueParserReporter()
+  private lazy val defaultGlobalForParser = ParseDriver()
+  private[sbt] def getGlobalForParser: ParseDriver = synchronized:
+    defaultGlobalForParser
```
Comment (adpi2) [inline buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: Object initialization is already thread-safe, so I don’t think these changes are actually fixing anything. Or am I missing something?

**Review by adpi2 — CHANGES_REQUESTED** [review]
I don't think this fixes the concurrency issue.

Looking at the stack trace, it seems that dotty is iterating the `System.properties` while another thread modifies them. Could it be that sbt, or a plugin modifies the `System.properties`?

**Comment by eed3si9n** [comment]
https://eed3si9n.com/december-adventure-2024/ has commentary on the whereabout of the thread.

File: buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala (hunk 1 above)
Comment (eed3si9n) [reply buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: defaultGlobalForParser is accessed by three places, including a thread we start as a background JIT warmup. I'm wrapping the access to synchronized so we don't give access to this variable concurrently.

Comment (adpi2) [reply buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: Its initialization and access are already thread safe.

Really I don't think the problem is here. It's rather that the `ParserDriver.<init>` which is called only once iterate over the `System.properties` while another thread is mutating those properties.

Comment (jtjeferreira) [reply buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: `java.util.Properties` is [supposedly](https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html) thread safe...
>  This class is thread-safe: multiple threads can share a single Properties object without the need for external synchronization

Comment (eed3si9n) [reply buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: We do 

sys.props.put("jna.nosys", "true")

in Main.scala and there might be more by users potentially. so maybe it's not safe to change and read at the same time?

Comment (jtjeferreira) [reply buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: Also [dotty.tools.dotc.config.WrappedProperties#systemProperties](https://github.com/scala/scala3/blob/1775d0be938fe78126600502f3c3c2e6d0429f9e/compiler/src/dotty/tools/dotc/config/WrappedProperties.scala#L24-L27) uses `System.getProperties.asScala.iterator` and maybe it should just use `System.getProperties.asScala`

Also `dotty.tools.dotc.config.PathResolver.Environment.javaBootClassPath` is `def javaBootClassPath: String = propOrElse("sun.boot.class.path", searchForBootClasspath)` and the 2nd argument is always evaluated.

But these are bugs for the scala repo...

Comment (jtjeferreira) [reply buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: And after some archeology: https://github.com/scala/scala/commit/f6859f28bb49193fde83e6020a6a89ce926a91e8

I will try to backport that commit to scala 3

Comment (eed3si9n) [reply buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: Added fdb7ca23841cf4db91545a206509542e12e65862 to wrap in `Retry(...)` instead.

**Review by adpi2 — APPROVED** [review]
LGTM

File: buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala (hunk 1 above)
Comment (eed3si9n) [reply buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala]: Thanks both for the review / detective work!

File: buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala [hunk 2]
```diff
@@ -132,7 +133,9 @@ private[sbt] object SbtParser:
   private[sbt] var scalacGlobalInitReporter: Option[ConsoleReporter] = None
 
   private[sbt] val globalReporter = UniqueParserReporter()
-  private[sbt] val defaultGlobalForParser = ParseDriver()
+  // Retry since Scala 3 compiler initialization can fail due to sys.props change
+  private[sbt] val defaultGlobalForParser: ParseDriver =
+    Retry(ParseDriver())
```
Comment (jtjeferreira) [inline buildfile/src/main/scala/sbt/internal/parser/SbtParser.scala:138]: this Retry did not work because it just retries `IOException`...

**Pull request merged**